File: src/GameImpl/MazeGenerator.fs
namespace BotRace.Game.Implementation.FSharp
open BotRace.Game
open _Maze
open _Cell

type RecursiveBacktrackingMazeGenerator() =
    
    let Directions = [| Direction.N; Direction.E; Direction.S; Direction.W |] 
    let rand = new System.Random()
    let Swap (a: _[]) x y = 
        let tmp = a.[x]
        a.[x] <- a.[y]
        a.[y] <- tmp
    let Shuffle a = 
        Array.iteri (fun i _ -> Swap a i (rand.Next(i, Array.length a))) a
        a

    let rec CarvePassageFrom (pos : _Position.T) (maze : _Maze.T)  = 
        Directions
        |> Shuffle 
        |> Array.iter
            (fun direction ->
                let target = pos |> _Position.At direction
                match 
                    target |> Exists maze
                    &&
                    (maze |> CellAt target) |> IsClosed with
                        | true ->
                            _Maze.Carve maze pos direction 
                            (maze |> CarvePassageFrom target) |> ignore
                        | false -> ())
        maze
    
    member this.Create size = 
        (this :> BotRace.Game.MazeGenerator).Create size

    interface BotRace.Game.MazeGenerator  with
        member this.Create size =
            size
            |> _Maze.Create 
            |> CarvePassageFrom (0, 0) 
            |> Implementation.FSharp.Maze
            :> BotRace.Game.Maze

File: src/Implementations/CSharp/GameImpl/MazeGenerator.cs
// http://weblog.jamisbuck.org/2010/12/27/maze-generation-recursive-backtracking    

// http://weblog.jamisbuck.org/2011/2/7/maze-generation-algorithm-recap.html

using BotRace.Game.Implementation.Extensions;
using IMaze = BotRace.Game.IMaze;

namespace BotRace.Game.Implementation
{
    public class RecursiveBacktrackingMazeGenerator : IMazeGenerator
    {
        public IMaze Create(int size)
        {
            var maze = Maze.ClosedGrid(size);

            var startPosition = new Position(0, 0);

            CarvePassageFrom(startPosition, maze);

            return maze;
        }

        private void CarvePassageFrom(Position pos, Maze maze)
        {
            var directions = new[] { Direction.N, Direction.E, Direction.S, Direction.W }.Shuffle();

            foreach (var direction in directions)
            {
                var targetPosition = (Position)pos.At(direction);

                if (maze.Exists(targetPosition) && maze.CellAt(targetPosition).IsClosed())
                {
                    maze.Carve(pos, direction);

                    CarvePassageFrom(targetPosition, maze);
                }
            }
        }
    }
}
